misc/multistream-select/src/protocol.rs

```diff
@@ -52,45 +50,32 @@ pub enum Version {
     /// is made to read from the `Negotiated` I/O stream).
     ///
-    /// A listener receiving a `V1Lazy` header will similarly delay sending
-    /// of the protocol confirmation.  Though typically the listener will need
-    /// to read the request data before sending its response, thus triggering
-    /// sending of the protocol confirmation, which, in absence of additional
-    /// buffering on lower layers will result in at least two response frames
-    /// to be sent.
-    ///
-    /// `V1Lazy` is specific to `rust-libp2p`: While the wire protocol
-    /// is identical to `V1`, delayed sending of protocol negotiation frames
-    /// is only safe under the following assumptions:
-    ///
-    ///   1. The dialer is assumed to always send the first multistream-select
-    ///      protocol message immediately after the multistream header, without
-    ///      first waiting for confirmation of that header. Since the listener
-    ///      delays sending the protocol confirmation, a deadlock situation may
-    ///      otherwise occurs that is only resolved by a timeout. This assumption
-    ///      is trivially satisfied if both peers support and use `V1Lazy`.
+    /// A listener will behave identically to `V1`. This ensures interoperability with `V1`.
+    /// Notably, it will immediately send the multistream header as well as the protocol
+    /// confirmation, resulting in multiple frames being sent on the underlying transport.
+    /// Nevertheless, if the listener supports the protocol that the dialer optimistically
+    /// settled on, it can be a 0-RTT negotiation.
     ///
-    ///   2. When nesting multiple protocol negotiations, the listener is either
-    ///      known to support all of the dialer's optimistically chosen protocols
-    ///      or there is no intermediate protocol without a payload and none of
-    ///      the protocol payloads has the potential for being mistaken for a
-    ///      multistream-select protocol message. This avoids rare edge-cases whereby
-    ///      the listener may not recognize upgrade boundaries and erroneously
-    ///      process a request despite not supporting one of the intermediate
-    ///      protocols that the dialer committed to. See [1] and [2].
+    /// > **Note**: `V1Lazy` is specific to `rust-libp2p`. The wire protocol is identical to `V1`
+    /// > and generally interoperable with peers only supporting `V1`. Nevertheless, there is a
+    /// > pitfall that is rarely encountered: When nesting multiple protocol negotiations, the
+    /// > listener should either be known to support all of the dialer's optimistically chosen
+    /// > protocols or there is must be no intermediate protocol without a payload and none of
+    /// > the protocol payloads must have the potential for being mistaken for a multistream-select
+    /// > protocol message. This avoids rare edge-cases whereby the listener may not recognize
```

Comment:
Let me know if I understand correctly:

If I create a protocol named `/foo` and the first message sent after opening a `/foo` substream is `/bar\nhello world`, then, if the listener doesn't support protocol `/foo`, it will misinterpret the payload as the dialer opening the protocol `/bar` and sending `hello world`.

Reply:
Yes, see https://github.com/multiformats/go-multistream/issues/20 for more details (that link is also in the code).